### PR TITLE
Allow configuration of the connection timeout to memcached

### PR DESCRIPTION
**Pull request by rcoonan-pd**
As currently implemented, the initial connection to memcached is using `:gen_tcp.connect`'s default timeout, which is `:infinity` (and thereby eventually hitting the lower-level TCP SYN retry limits which total 130 seconds). We have experienced issues in our systems where a memcached host became unresponsive, and the large timeout described above caused the system to enter a problematic state before properly realizing the host was gone and moving onto a different host. As such, we would like the ability to configure the timeout for the connection.

In this implementation, it can be passed in as an option to Connection. ~However, given the ubiquity of Connection throughout the library, I also added a way to just configure the default timeout value so that every call in doesn't have to be passing the timeout option down.~

Happy to take any feedback or make any proposed changes, @ananthakumaran

### CODE REVIEW

File: lib/memcache/connection.ex [hunk 1]
```diff
@@ -47,6 +47,10 @@ defmodule Memcache.Connection do
     to be used in case of connection failure. Defaults to `500`
   * `:backoff_max` - (integer) maximum allowed interval between two
     connection attempt. Defaults to `30_000`
+  * `:timeout` - (integer) connection timeout (in milliseconds) at
```
Comment (ananthakumaran) [inline lib/memcache/connection.ex]: could we call it `connect_timeout` to differentiate with send_timeout etc

File: lib/memcache/connection.ex [hunk 2]
```diff
@@ -71,7 +75,10 @@ defmodule Memcache.Connection do
   ]
 
   defp with_defaults(opts) do
+    default_timeout = Application.get_env(:memcachex, :default_timeout, :infinity)
```
Comment (ananthakumaran) [inline lib/memcache/connection.ex]: I prefer not to touch application env especially if the application is not singleton. None of the other configs are controlled via application env as well. I would suggest to add it to the `@default_opts`

**Review by ananthakumaran — CHANGES_REQUESTED** [review]

File: lib/memcache/connection.ex (hunk 1 above)
Comment (rcoonan-pd) [reply lib/memcache/connection.ex]: Sure, yeah, makes sense 👍

File: lib/memcache/connection.ex (hunk 2 above)
Comment (rcoonan-pd) [reply lib/memcache/connection.ex]: Hm, yes, I recognize that this was the only place using the Application env. Given that the `@default_opts` are hardcoded, though, is your preference then that there is no way to globally configure the timeout? Rather, that a timeout should be passed in the opts with every call (that wishes to not use infinity)?

Comment (ananthakumaran) [reply lib/memcache/connection.ex]: > Rather, that a timeout should be passed in the opts with every call (that wishes to not use infinity)?

It's only required for start_link on Memcache/Connection module. So I expect it to be needed at one place (wherever the memcache supervisor hierarchy is configured)

Comment (rcoonan-pd) [reply lib/memcache/connection.ex]: Alright, makes sense. I'll put together some changes early next week, and ping you when it's ready to be looked at again. Thanks for the quick review!

**Comment by rcoonan-pd** [comment]
@ananthakumaran Alright, I have updated the PR according to your feedback.

**Review by ananthakumaran — APPROVED** [review]

**Pull request merged**